diff --git a/node_modules/cypress/types/cypress.d.ts b/node_modules/cypress/types/cypress.d.ts
index e231155..1be9711 100644
--- a/node_modules/cypress/types/cypress.d.ts
+++ b/node_modules/cypress/types/cypress.d.ts
@@ -2115,12 +2115,17 @@ declare namespace Cypress {
      */
     spread(fn: (...args: any[]) => void): Chainable<Subject>
 
+
     /**
      * Run a task in Node via the plugins file.
      *
      * @see https://on.cypress.io/api/task
      */
-    task<S = unknown>(event: string, arg?: any, options?: Partial<Loggable & Timeoutable>): Chainable<S>
+    task<K extends keyof Tasks & string>(
+      event: K,
+      arg: Parameters<Tasks[K]>[0],
+      options?: Partial<Loggable & Timeoutable>
+    ): Chainable<Awaited<ReturnType<Tasks[K]>>>;
 
     /**
      * Enables you to work with the subject yielded from the previous command.