diff --git a/node_modules/node-calls-python/index.js b/node_modules/node-calls-python/index.js
index 07a62cd..bbecc5e 100644
--- a/node_modules/node-calls-python/index.js
+++ b/node_modules/node-calls-python/index.js
@@ -1,7 +1,7 @@
 const { execSync } = require("child_process");
 const fs = require("fs");
 const path = require("path");
-const nodecallspython = require("./build/Release/nodecallspython");
+const nodecallspython = require("bindings")("nodecallspython");
 
 class Interpreter
 {
